Modules/main.c

```diff
@@ -244,6 +468,12 @@ pymain_run_command(wchar_t *command)
         return pymain_exit_err_print();
     }
 
+    // Only perform auto-dedent if the string starts with a newline
+    /*if (wcsncmp(command, L"\n", 1) == 0) {                         */
+    /*    // Should the input be modified here with the Python C-API?*/
+    /*    unicode = _unicode_dedent(unicode);                        */
+    /*}                                                              */
+
     bytes = PyUnicode_AsUTF8String(unicode);
```

Comment:
You should directly make changes to this `bytes` object at this point. (of course after checking if it is null)

1. Do not read anything before `PySys_Audit`.
2. `bytes` is `UTF-8` encoded, much easier to handle than `wchar_t`.
3. This preserves the code structure.

Reply:
Then you can implement your dedent algo, and replace the `bytes`.

Also, don't forget to check the ref count after.

Reply:
I've replaced the old implementations with one that works directly on `char *`. Much of the implementation was ported from the old wchar_t version, but it is a little simpler. I'm using `PyBytes_FromStringAndSize` to do memory allocation.

Should I be avoiding mutating the input bytes objects (i.e. allocating new memory for the dedented data)? Or should I just clobber the user input and avoid the `strtok` so new memory doesn't need to be allocated? (Although I think having at least one temporary buffer does make the algorithm a lot simpler).

Reply:
> (i.e. allocating new memory for the dedented data)

allocating is fine, use `Py_Malloc` and `Py_Free`




> Much of the implementation was ported from the old wchar_t version

I think you should use a handwritten implementation like
1. count before allocate, early exit if already met line with zero indent.
```
char * data = PyBytes_AsString(bytes);


int indent_count = ....;
char c;
while (c = *data++){
if (c == '\n') ...
if (c == ' ')...
// ...
}
```

2. if dedent is really needed, allocate a new `char*`, write to it.
3. create a new `PyByteObject` from `char*`.
4. free the `char*`, and old `bytes`


I think this is enough for a demo. But it's better to use the internal API: `_PyBytesWriter_*` if you want to make it production quality

Reply:
@sunmy2019 I gave a handwritten implementation a shot, including using `_PyBytesWriter_*`. The implementation should now cover tabs as well as spaces. I kept the old implementation commented out for now. I will remove it if you like the new implementation (or think it can be fixed into something mergable).

One thing I think might be improved is the way I'm handling grabing a single line at a time. I wasn't able to use the `(c = *data++)` idiom because I need to execute the loop body if the current char is `\n` or if we reach the NULL byte at the end of the string.

I'm also not 100% sure if my usage of `_PyBytesWriter_` is correct. I tried to follow other places it is used in the code.

I took your suggestion and I just return the original bytes if no common leading spaces are found, and only ran the `Py_DECREF` if the output wasn't the same pointer as the input to dedent. Not sure if that is confusing memory management or not, but it seemed reasonable.